File: android/app/src/main/java/com/github/bfollon/farmaciasdeguardiaensegovia/services/NetworkMonitor.kt
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.bfollon.farmaciasdeguardiaensegovia.services

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities

/**
 * Network monitoring utility to check online/offline state
 * Simple wrapper around Android's ConnectivityManager
 */
object NetworkMonitor {
    
    private var connectivityManager: ConnectivityManager? = null
    
    /**
     * Initialize the network monitor with application context
     * Should be called once from MainActivity.onCreate()
     */
    fun initialize(context: Context) {
        connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        DebugConfig.debugPrint("NetworkMonitor: Initialized")
    }
    
    /**
     * Check if the device has an active internet connection
     * 
     * @return true if device has validated internet access, false otherwise
     */
    fun isOnline(): Boolean {
        val cm = connectivityManager
        if (cm == null) {
            DebugConfig.debugWarn("NetworkMonitor: Not initialized, assuming offline")
            return false
        }
        
        return try {
            val network = cm.activeNetwork
            if (network == null) {
                DebugConfig.debugPrint("NetworkMonitor: No active network")
                return false
            }
            
            val capabilities = cm.getNetworkCapabilities(network)
            if (capabilities == null) {
                DebugConfig.debugPrint("NetworkMonitor: No network capabilities")
                return false
            }
            
            // Check for actual internet capability (not just network connection)
            val hasInternet = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            val isValidated = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
            
            val online = hasInternet && isValidated
            
            DebugConfig.debugPrint("NetworkMonitor: Online status = $online (internet=$hasInternet, validated=$isValidated)")
            online
            
        } catch (e: Exception) {
            DebugConfig.debugError("NetworkMonitor: Error checking network state", e)
            false
        }
    }
    
    /**
     * Get a human-readable description of the current network state
     * Useful for debugging and UI display
     */
    fun getNetworkStateDescription(): String {
        val cm = connectivityManager ?: return "No inicializado"
        
        return try {
            val network = cm.activeNetwork ?: return "Sin conexión"
            val capabilities = cm.getNetworkCapabilities(network) ?: return "Sin capacidades de red"
            
            val transportType = when {
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "WiFi"
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "Datos móviles"
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "Ethernet"
                else -> "Desconocido"
            }
            
            val validated = if (capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)) {
                "conectado"
            } else {
                "sin internet"
            }
            
            "$transportType ($validated)"
            
        } catch (e: Exception) {
            DebugConfig.debugError("NetworkMonitor: Error getting network description", e)
            "Error al comprobar red"
        }
    }
}

File: ios/FarmaciasDeGuardiaEnSegovia/Services/NetworkMonitor.swift
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import Foundation
import Network
import Combine

/// Network monitoring utility to check online/offline state
/// Uses Apple's Network framework (NWPathMonitor) for real-time connectivity detection
/// Observable for SwiftUI integration
class NetworkMonitor: ObservableObject {

    /// Shared singleton instance
    static let shared = NetworkMonitor()

    /// Published property for SwiftUI observation - true if online, false if offline
    @Published private(set) var isOnline: Bool = true

    private let monitor: NWPathMonitor
    private let queue = DispatchQueue(label: "NetworkMonitor")

    private init() {
        monitor = NWPathMonitor()

        monitor.pathUpdateHandler = { [weak self] path in
            let wasOnline = self?.isOnline ?? true
            let nowOnline = path.status == .satisfied

            DispatchQueue.main.async {
                self?.isOnline = nowOnline

                // Log status changes
                if wasOnline != nowOnline {
                    if nowOnline {
                        DebugConfig.debugPrint("📡 NetworkMonitor: Connection restored")
                    } else {
                        DebugConfig.debugPrint("📡 NetworkMonitor: Connection lost")
                    }
                }

                // Detailed logging
                self?.logNetworkState(path)
            }
        }

        monitor.start(queue: queue)
        DebugConfig.debugPrint("📡 NetworkMonitor: Initialized and monitoring")
    }

    deinit {
        monitor.cancel()
    }

    /// Get a human-readable description of the current network state
    /// Useful for debugging and UI display
    func getNetworkStateDescription() -> String {
        let path = monitor.currentPath

        guard path.status == .satisfied else {
            return "Sin conexión"
        }

        let transportType: String
        if path.usesInterfaceType(.wifi) {
            transportType = "WiFi"
        } else if path.usesInterfaceType(.cellular) {
            transportType = "Datos móviles"
        } else if path.usesInterfaceType(.wiredEthernet) {
            transportType = "Ethernet"
        } else {
            transportType = "Desconocido"
        }

        let validated = path.status == .satisfied ? "conectado" : "sin internet"

        return "\(transportType) (\(validated))"
    }

    /// Log detailed network state for debugging
    private func logNetworkState(_ path: NWPath) {
        let status = path.status == .satisfied ? "Online" : "Offline"
        let interface = getInterfaceType(path)
        let expensive = path.isExpensive ? "expensive" : "not expensive"
        let constrained = path.isConstrained ? "constrained" : "not constrained"

        DebugConfig.debugPrint("📡 NetworkMonitor: Status=\(status), Interface=\(interface), \(expensive), \(constrained)")
    }

    /// Get the current interface type
    private func getInterfaceType(_ path: NWPath) -> String {
        if path.usesInterfaceType(.wifi) {
            return "WiFi"
        } else if path.usesInterfaceType(.cellular) {
            return "Cellular"
        } else if path.usesInterfaceType(.wiredEthernet) {
            return "Ethernet"
        } else if path.usesInterfaceType(.loopback) {
            return "Loopback"
        } else {
            return "Unknown"
        }
    }
}
